src/anomalib/training/trainer.py

```diff
@@ -0,0 +1,96 @@
+
+log = logging.getLogger(__name__)
+# warnings to ignore in trainer
+warnings.filterwarnings(
+    "ignore", message="torch.distributed.reduce_op is deprecated, please use torch.distributed.ReduceOp instead"
+)
+
+
+class AnomalibTrainer(Trainer):
+    """Anomalib trainer.
+
+
+    Note:
+        Refer to PyTorch Lightning's Trainer for a list of parameters for details on other Trainer parameters.
+
+    Args:
+        threshold_method (ThresholdMethod): Thresholding method for normalizer.
+        normalization_method (NormalizationMethod): Normalization method
+        manual_image_threshold (Optional[float]): If threshold method is manual, this needs to be set. Defaults to None.
+        manual_pixel_threshold (Optional[float]): If threshold method is manual, this needs to be set. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        threshold_method: ThresholdMethod = ThresholdMethod.ADAPTIVE,
+        normalization_method: NormalizationMethod = NormalizationMethod.MIN_MAX,
+        manual_image_threshold: Optional[float] = None,
+        manual_pixel_threshold: Optional[float] = None,
+        image_metrics: list[str] | None = None,
+        pixel_metrics: list[str] | None = None,
+        task_type: TaskType = TaskType.SEGMENTATION,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+
+        self.lightning_module: AnomalyModule  # for mypy
+
+        self.fit_loop = AnomalibFitLoop(
+            min_epochs=kwargs.get("min_epochs", 0), max_epochs=kwargs.get("max_epochs", None)
+        )
+        self.validate_loop = AnomalibValidationLoop()
+        self.test_loop = AnomalibTestLoop()
+        self.predict_loop = AnomalibPredictionLoop()
+
+        self.task_type = task_type
+
+        # TODO: configure these from the config
+        self.post_processor = PostProcessor(
+            threshold_method=threshold_method,
+            manual_image_threshold=manual_image_threshold,
+            manual_pixel_threshold=manual_pixel_threshold,
+        )
+        self.normalizer = Normalizer(normalization_method=normalization_method)
+        self.metrics_manager = MetricsManager(image_metrics=image_metrics, pixel_metrics=pixel_metrics)
+
+    def _call_setup_hook(self) -> None:
+        """Override the setup hook to call setup for required anomalib classes.
+
+        Ensures that the necessary attributes are added to the model before callbacks are called. The majority of the
+        code is same as the base class. Add custom setup only between the commented block.
+        """
+        assert self.state.fn is not None
+        fn = self.state.fn
+
+        self.strategy.barrier("pre_setup")
+
+        if self.datamodule is not None:
+            self._call_lightning_datamodule_hook("setup", stage=fn)
+
+        # Setup required classes before callbacks and lightning module
+        self.post_processor.setup(self.lightning_module)
+        self.normalizer.setup()
```

Comment:
We don't have access to the `lightning_module` until it is passed to `fot`, `test`, etc. I'll move the ones that do not depend on the `lightning_module` to constructor. There might be a better approach than calling setup. I'll investigate.

Reply:
If we keep the metrics as attributes of the `MetricsManager`, and the thresholds as attributes of the `PostProcessor`, would we still need access to the lightning module? If not, this would be another advantage of keeping the metrics in the 'manager' classes instead of the model.